### ISSUE TITLE
[engine-handlebars] Included pattern will not render default data if called inside a block

### ISSUE THREAD

**Issue by mrkmiller** [pattern-lab/patternlab-node#962]
I am using Pattern Lab Node `v3.0.0-beta.1` on `Mac`, with Node `v8.11.2`, using a `Custom` Edition.

```
"@pattern-lab/core": "^3.0.0-beta.1",
"@pattern-lab/engine-handlebars": "^2.0.0-beta.0",
"@pattern-lab/uikit-workshop": "^1.0.0-beta.0",
```

##### Expected Behavior
When including a pattern inside another pattern such as an atom within a molecule, if including that pattern inside of a block `{{# }}`, the data set originally in the global `data.json` should be used if the surrounding object doesn't redefine the data.

This is how the mustache engine works.

##### Actual Behavior
The atom will display the html but without any data.

##### Steps to Reproduce
1. Add default data to the global `data.json` file.
```
{
  "btnText": "Default Button",
  "parentObject": {
    "text": "Inside parent object"
  }
}
```
2. Create a `button.hbs` atom.
```
<button>{{ btnText }}</button>
```
3. Create a `parent.hbs` molecule.
```
Correctly Renders button as "Default Button"
{{> atoms-button }}

{{# parentObject }}
  {{ text }}
  
  Incorrectly Renders button as ""
  {{> atoms-button }}
{{/ parentObject}}
```

**Comment by david-moldenhauer** [comment]
@mrkmiller I believe this is a context issue. The context should be one level up:
---------------------------------------------------------
Correctly Renders button as "Default Button"
{{> atoms-button }}

{{# parentObject }}
  {{ text }}
  
  Incorrectly Renders button as ""
  {{> atoms-button ../this}}
{{/ parentObject}}

**Comment by mrkmiller** [comment]
Thanks for the quick response. Passing an explicit context will fix things when viewing the Molecule by itself, but it then breaks things higher up in a Page for example when trying  If the explicit context is set, then there is no way to override the button text in page data inside a loop. For example, if the molecule is now rendered multiple times in a list with each item needing a custom button label, there is no way to do that.

I'm converting an existing Pattern Lab 1 site so am coming from a mustache mindset. If this is simply the way handlebars works then it is something to consider with the proposed move to handlebars as the default engine. So far my experience moving from mustache to handlebars is that it has completely broken the data inheritance model and the site display.

In short, I've found that the way data inheritance works between mustache and handlebars to be completely different.

**Comment by bradfrost** [comment]
Hey @mrkmiller, thanks for sharing the issue. I very much am coming from a Mustache perspective, so color me interested. I unfortunately don't have solutions here, just chiming in to say that we're all really invested in making the transition from Mustache to Handlebars as smooth as possible. /cc @EvanLovely @bmuenzenmeyer

**Comment by david-moldenhauer** [comment]
@mrkmiller I've been working with HBS for a while now. let me give you a few examples that might help with your scenario. first of all in your example the context would move into the parentObject for anything that is in the {{# parentObject }} block meaning anything in the block would be passed only "text". Now within the same template you can move within the context with "../" syntax. Within the {{> atoms-button }} however the context that it gets is only the passed object with it's subelements.
You can however save a context within your hbs files by using hbs-helpers.  
 If you would use these helpers (https://github.com/david-moldenhauer/plugin-node-handlebars-helpers) you could for example do the following:
Add default data to the global data.json file.
```
{
  "global":{
    "btnText": "Default Button",
    "parentObject": [{
      "text": "Inside parent object"
    }, {
      "text": "Inside parent object 2"
    }]
  }
}
```
Create a button.hbs atom.
```
{{#set}}{{{file "./source/_data/data.json"}}}{{/set}}
<button>{{eitherOr btnText global.btnText (lorem-ipsum (add 1 (random 2)) 'words' 'plain')}}</button>
```
the eitherOr helper will choose the first true argument passed so if btnText does not exist it will fall back on the global.btnText which will be available as it is injected into the context by the "set" and "file" helper (set takes a json). However if the global.btnText should not be available it will fallback to 1-3 words of lorem-ipsum.
 
Create a parent.hbs molecule.
Correctly Renders button as "Default Button"
```
{{> atoms-button }}

{{# parentObject }}
  {{ text }}
  
  correctly Renders button as global.btnText
  {{> atoms-button }}
  correctly Renders button as parentObject[currentindex].text
  {{> atoms-button btnText=text}}
{{/ parentObject}}
```

hope this helps.

**Comment by engelfrost** [comment]
TL;DR

Use `{{#if}}`, it keeps the context

---

I'm in the process of migrating from Mustache to Handlebars. I have found the inheritance in Mustache to be both helpful and annoying. Handlebars lets me handle this complexity better. 

I don't use "pure" blocks in Handlebars. I've changed every `{{#foo}}` in Mustache to either `{{#if foo}}` or `{{#each foo}}` in Handlebars. 

[And `{{#if}}` does not change the context.](http://handlebarsjs.com/#paths)

So in the example from @mrkmiller the following should work as expected: 
```json
{
  "btnText": "Default Button",
  "parentObject": {
    "text": "Inside parent object"
  }
}
```
```hbs
{{#if parentObject}}
  {{ text }}
  {{> atoms-button }}
{{/if}}
```

If I want to change the context and be explicit about it I use `{{#with}}`. I use this for "namespacing" global values:

`_data/items.json`
```json
{
  "items": [ ... ]
}

`_data/navigation.json`
```json
{
  "navigation": {
    "items": [ ... ]
  }
}
``` 

```hbs
{{#if navigation}}
  {{#with navigation}}
    {{#each items}}
      <!-- These items are from navigation.json. I can easily override them in pseudo patterns. -->
    {{/each}} 
  {{/with}}
{{/if}}
```

**Comment by bradfrost** [comment]
Hey @engelfrost,
Thanks very much for this handy breakdown! I think this is a great overview for anyone making the transition from Mustache to Handlebars. 

I'm going to close this issue out, but let us know if there's still more to cover here.
